### ISSUE TITLE
init, props, and messaging children

### ISSUE THREAD

**Issue by uniphil** [uniphil/spindle#8]
Right now, `handleProps : Props -> Model -> Update` is used alongside `update` to change the model, and it fills the need for:
1. an `init` function that can operate on `props`
2. `props` that a child cares about that may change
3. a poor kind of "messaging" from parent -> child via changing props.

**1.**, it's slightly awkward, with possibly surprising effects. If all you want is `init`, you probably don't want to have to worry about `init` re-running at any time just because a parent changed a prop.

**2.** works ok. It might be slightly anti-react, since you have to jump through a (very low) hoop to access props from the view. If that becomes an annoying level of usually-useless boilerplate, `props` could become available in `view`. I don't think that's ever better, but it's more convenient and at the end of the day, that counts.

**3.** I'm going to ignore for now. I don't think there's anything special or different from React. The main use-case I have for controlling children with something event-like is resetting a form, and there might be better ways to handle that: unmounting and re-mounting the component will do the trick but it's not nice to actually write that. I think there can be something nice to make that happen though. In fact, having a couple escape hatches for managing children might be nice, eg if a component gets moved in the DOM but we don't want react to kill it and lose its state and start over. Later problems though.

So I'm leaning slightly toward re-adding `init` with the same signature as `handleProps`. That function can then be renamed to `propsUpdate`, which I like more. Props _miiiiiiight_ be made available in more places like `view` and `subscriptions` if they prove to be useful shortcuts.

It's hard to let the tiny api slip and grow a bit :cry: but ohhhhhhhh well.

**Comment by uniphil** [comment]
also, since the component state is now _always_ initialized to just `Model()`, `init` will be optional on only have to be busted out when you want to grab stuff from props on mount.

splitting `init` out also maps pretty nicely to React's lifecycle methods, since `componentWillReceiveProps` is not called on mount, so currently `spindle` calls `handleProps` in two places.

**Comment by uniphil** [comment]
`examples/constrained-counters/` and `examples/wrap/` both get worse if `init` is simply added back in -- in both cases, they're just caching props onto the model, so they want to do that during `init` and during `propsUpdate`. It's not cool to write the function out twice, and

``` js
const init = (...args) =>
  propsUpdate(...args);
```

is not that nice of a pattern either.

So I think either spindle needs to
1. Stick to the current way, with `handleProps` only and no special `init`, and maybe keep `props` only accessible from `handleProps`, or
2. Add `init` but immediately go further and make `props` available to `update`, `view`, and `subscription`, so that simply caching stuff from props onto the model is not a pattern at all. Both `constrained-counters` and `wrap` won't need `init` _or_ `propsUpdate`.

**2.** is the sad route but ultimately probably _much_ more enjoyable to write apps with, I think. One nice thing about it is that `children` doesn't have to be special any more. It's generally a little more react-friendly.

But this might also break everything. Or stuff might be broken anyway haha. If `update` computes and saves a value that depends on `props`, then we won't (and currenly don't?) have a way to tell `update` to re-compute that. Arg!

**Comment by uniphil** [comment]
If there's a way to make a kind of subscription to `props` that feeds into `update` via some Msg, and if that's the only way to access `props` in `update`, then maybe all (most of?) these problems go away.

**Comment by uniphil** [comment]
The other semi-obvious (and probably right?) solution is to replace `Model` with `init`, being a function

```
props -> Update
```

(w/ 2nd param `BoundMsg` in case the the function wants to do `cmds` in `Update`)

This would make the normal case where we're just using `Model = Record({` slightly more syntax-heavy, but
- not forcing `Record` as the only kind of `model` is kind of nice
- `const Model = Record({...})` returning a function is a bit weird unless you're deep into immutable fu, and making `init` a little more "normal" is nice
- I think it solves all the problems? and makes sense?

Where we're just initializing static state currently, the code becomes:

```
const Model = Record({
  a: 1,
  b: 2,
});

const init = () => Update({ model: Model() });
```

basically, `Model` becomes something that components might want, but don't need to tell Spindle about any more.

**Comment by uniphil** [comment]
done.

phew!
